.github/workflows/test-pull-request-java.yml

```diff
@@ -1,63 +1,87 @@
+          do
+            echo "working on $changed_folder"
+            smoke_test=""
+            current_folder=$changed_folder
+            while [ -z "$smoke_test" ] && [ "$current_folder" != "." ]
+            do
+              echo "--smoke_test is $smoke_test"
+              echo "--current_folder is $current_folder"
+              if [[ -s $current_folder/smoke_test ]]; then
+                smoke_test="$current_folder/smoke_test"
+              else
+                current_folder=$(dirname $current_folder)
+              fi
+            done
+            if [ "$current_folder" != "." ]; then
+              smoke_test_folders="$smoke_test_folders\"$smoke_test\"\n"
+            fi
+          done <<<"$changed_java_folders"
+
+          # Once the smoke_test_folders are all found, we de duplicate them (`uniq`), remove extra empty lines (`\S`), and format them into a string
+          # we end up with, for example, smoke_test_folders='"java/akka/smoke_test","java/kotlin/smoke_test","java/application-insights/smoke_test"'
+          # see: https://stackoverflow.com/a/8721550/24069 for usage of paste
+          smoke_test_folders=$( echo -e $smoke_test_folders | uniq  | grep "\S" | paste -s -d, -)
+
+          echo "These are the Java folders where we found changes for this PR : $changed_java_folders" >> $GITHUB_STEP_SUMMARY
+          echo "These are the Java folders where we found changes AND the presence of smoke tests for this PR : $smoke_test_folders" >> $GITHUB_STEP_SUMMARY
+          echo "matrix={\"java_sample_folder\":[$smoke_test_folders]}" >> $GITHUB_OUTPUT
+          echo "This is how the matrix will look like: $(cat $GITHUB_OUTPUT)"
+
   smoke:
+    needs: prepare
     name: Smoke Tests
     runs-on: ubuntu-latest
+    strategy:
+      matrix: ${{ fromJSON(needs.prepare.outputs.matrix) }}
+
     steps:
-    - name: Setup Go
-      uses: actions/setup-go@v4
-      with:
-        go-version: 1.18.x
-
-    - name: Checkout
-      uses: actions/checkout@v4
-
-    - name: Cache local Gradle repository
-      uses: actions/cache@v3
-      with:
-        path: |
-          ~/.gradle/caches
-          ~/.gradle/wrapper
-        key: ${{ runner.os }}-gradle-${{ hashFiles('java/**/*.gradle*', 'java/**/gradle-wrapper.properties') }}
-        restore-keys: |
-          ${{ runner.os }}-gradle-
-
-    - name: Cache local Maven repository
-      uses: actions/cache@v3
-      with:
-        path: |
-          ~/.m2/repository
-          ~/.m2/wrapper
-        key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml', 'java/**/maven-wrapper.properties') }}
-        restore-keys: |
-          ${{ runner.os }}-maven-
-
-    - name: Setup Directories
-      run: |
-        #!/usr/bin/env bash
-        set -euo pipefail
-
-        mkdir -p ~/.m2/repository
-        mkdir -p ~/.m2/repository/wrapper/dists
-        mkdir -p ~/.gradle/caches
-        mkdir -p ~/.gradle/wrapper/dists
-        chmod -R 775 ~/.m2
-        chmod -R 775 ~/.gradle
```

Comment:
This part @anthonydahanne

Reply:
@dmikusa and I agreed on letting caching go for now; ta favor simplicity.

If things go south, I'll re implement it